src/ucs/arch/x86_64/cpu.c

```diff
@@ -94,37 +102,47 @@ static double ucs_x86_tsc_freq_from_cpu_model()
         max_ghz = ucs_max(max_ghz, ghz);
         if (max_ghz != ghz) {
             warn = 1;
+            break;
         }
     }
     fclose(f);
 
     if (warn) {
-        ucs_debug("Conflicting CPU frequencies detected, using fallback timer");
-        ucs_arch_x86_enable_rdtsc = 0;
+        level = ucs_global_opts.warn_inv_tsc ? UCS_LOG_LEVEL_WARN :
+                                               UCS_LOG_LEVEL_DEBUG;
+        ucs_log(level,
+                "Conflicting CPU frequencies detected, using fallback timer");
+
+        return -1;
     }
-    return max_ghz * 1e9;
+
+    *result = max_ghz * 1e9;
+
+    return 0;
 }
 
 double ucs_arch_get_clocks_per_sec()
 {
-    double result;
+    double freq;
+    int ret;
 
-    ucs_x86_check_invariant_tsc();
+    ret = ucs_x86_tsc_freq_from_cpu_model(&freq);
 
-    if (ucs_arch_x86_enable_rdtsc) {
-        /* First, try to find the information in the model name string (will work on Intel) */
-        result = ucs_x86_tsc_freq_from_cpu_model();
-        if (result == 0) {
-            /* Read clock speed from cpuinfo */
-            result = ucs_get_cpuinfo_clock_freq("cpu MHz", 1e6);
-        }
-        if ((result > 0.0) && ucs_arch_x86_enable_rdtsc) {
-            /* using rdtsc */
-            return result;
-        }
+    if (ret) {
+        goto fallback;
+    }
+
+    if (freq == 0) {
+        /* Read clock speed from cpuinfo */
+        freq = ucs_get_cpuinfo_clock_freq("cpu MHz", 1e6);
+    }
+
+    if (freq > 0.0) {
+        /* using rdtsc */
+        return freq;
     }
 
-    /* fallback */
+fallback:
```

Comment:
seems there is a mismatch: the timer frequency depends on the result of ucs_get_cpuinfo_clock_freq(), while the decision of whether to call ucs_arch_generic_read_hres_clock() depends only on the result of ucs_x86_tsc_freq_from_cpu_model()

Reply:
`ucs_get_cpuinfo_clock_freq` does not change the state of `ucs_arch_x86_enable_rdtsc`.

Reply:
this is the problem, ucs_get_cpuinfo_clock_freq() affects ucs_get_cpu_clocks_per_sec(), which is clock frequency, but does not affect ucs_arch_x86_enable_rdtsc, which is clock source. looks like there could be mismatch between clock source and assumed clock frequency?

Reply:
was not this an intention? it worked the same way before, `ucs_get_cpuinfo_clock_freq()` did not change `ucs_arch_x86_enable_rdtsc`, we just took the max value. 
Results should be consistent anyway, because we always use the same clock  and its frequency values from the beginning.